contracts/zero-ex/contracts/src/transformers/PayTakerTransformer.sol

```diff
@@ -0,0 +1,89 @@
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.6.5;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-utils/contracts/src/v06/errors/LibRichErrorsV06.sol";
+import "@0x/contracts-utils/contracts/src/v06/LibSafeMathV06.sol";
+import "@0x/contracts-erc20/contracts/src/v06/IERC20TokenV06.sol";
+import "@0x/contracts-erc20/contracts/src/v06/LibERC20TokenV06.sol";
+import "../errors/LibTransformERC20RichErrors.sol";
+import "./Transformer.sol";
+import "./LibERC20Transformer.sol";
+
+
+/// @dev A transformer that transfers tokens to the taker.
+contract PayTakerTransformer is
+    Transformer
+{
+    // solhint-disable no-empty-blocks
+
+    /// @dev Transform data to ABI-encode and pass into `transform()`.
+    struct TransformData {
+        // The tokens to transfer to the taker.
+        IERC20TokenV06[] tokens;
+        // Amount of each token in `tokens` to transfer to the taker.
+        // `uint(-1)` will transfer the entire balance.
+        uint256[] amounts;
+    }
+
+    using LibRichErrorsV06 for bytes;
+    using LibSafeMathV06 for uint256;
+    using LibERC20Transformer for IERC20TokenV06;
+
+    /// @dev Create this contract.
+    /// @param deploymentNonce_ The nonce of the deployer when deploying this contract.
+    /// @dev Construct the transformer and store the WETH address in an immutable.
+    constructor(uint256 deploymentNonce_)
+        public
+        Transformer(deploymentNonce_)
+    {}
+
+    /// @dev Forwards tokens to the taker.
+    /// @param taker The taker address (caller of `TransformERC20.transformERC20()`).
+    /// @param data_ ABI-encoded `TransformData`, indicating which tokens to transfer.
+    /// @return rlpDeploymentNonce RLP-encoded deployment nonce of the deployer
+    ///         when this transformer was deployed. This is used to verify that
+    ///         this transformer was deployed by a trusted contract.
+    function transform(
+        bytes32, // callDataHash,
+        address payable taker,
+        bytes calldata data_
+    )
+        external
+        override
+        returns (bytes memory rlpDeploymentNonce)
+    {
+        TransformData memory data = abi.decode(data_, (TransformData));
+
+        // Transfer tokens directly to the taker.
+        for (uint256 i = 0; i < data.tokens.length; ++i) {
+            // The `amounts` array can be shorter than the `tokens` array.
+            // Missing elements are treated as `uint256(-1)`.
+            uint256 amount = data.amounts.length > i ? data.amounts[i] : uint256(-1);
+            if (amount == uint256(-1)) {
+                amount = data.tokens[i].getTokenBalanceOf(address(this));
+            }
+            if (amount != 0) {
```

Comment:
I wonder if we should error out when `amount == 0` 🤔. Is there a use case for this?

Reply:
Yeah. The `PayTakerTransformer` is also used to refund unspent protocol fees, which could end up being `0` if they were all used.